### ISSUE TITLE
stream.rotate does not take account the component azimuth

### ISSUE THREAD

**Issue by rdno** [obspy/obspy#1604]
Hi,

Rotated seismograms (`NE->RT`) produced using `obspy` and `SAC` were different for some stations. I noticed that in my data component azimuths do not always reflect the station name. In the example below, north component has an azimuth of 270. From what I understand, obspy does not take the component azimuth into account while performing the rotation. On the other hand, SAC reads this data before rotation. I tested this by creating a new set of seismograms with different component azimuths and rotating them.

Does obspy have a routine for this or am I missing something?

Obspy version: 1.0.2

---

Example for sac:
```
 SEISMIC ANALYSIS CODE [11/11/2013 (Version 101.6a)]
 Copyright 1995 Regents of the University of California

SAC> r ./TRIS.IU.BHN.obs.sac ./TRIS.IU.BHE.obs.sac ./TRIS.IU.BHN.mod.obs.sac ./TRIS.IU.BHE.mod.obs.sac
SAC> lh cmpaz


  FILE: ./TRIS.IU.BHN.obs.sac - 1
 ---------------------------

     cmpaz = 2.700000e+02

  FILE: ./TRIS.IU.BHE.obs.sac - 2
 ---------------------------

     cmpaz = 0.000000e+00

  FILE: ./TRIS.IU.BHN.mod.obs.sac - 3
 -------------------------------

     cmpaz = 0.000000e+00

  FILE: ./TRIS.IU.BHE.mod.obs.sac - 4
 -------------------------------

     cmpaz = 9.000000e+01
SAC> rotate to gcp
SAC> color on inc
SAC> p1
SAC> w ./TRIS.IU.BHR.obs.sac ./TRIS.IU.BHT.obs.sac ./TRIS.IU.BHR.mod.obs.sac ./TRIS.IU.BHT.mod.obs.sac
```
Plot:
![sac_rt](https://cloud.githubusercontent.com/assets/77037/20710681/ccb2b9ce-b63b-11e6-910b-8f3f7b3256df.png)

Plotting the rotated seismograms of original versions reveals the discrepancy between obspy and SAC (I read the back azimuth value from sac headers):

```python
import obspy
import matplotlib.pyplot as plt

ne = obspy.read("./TRIS.IU.BHN.obs.sac")
ne.append(obspy.read("./TRIS.IU.BHE.obs.sac")[0])

rt = obspy.read("./TRIS.IU.BHR.obs.sac")
rt.append(obspy.read("./TRIS.IU.BHT.obs.sac")[0])

baz = rt[0].stats.sac['baz']

ne.rotate("NE->RT", back_azimuth=baz)

fig, axs = plt.subplots(nrows=2)

for ax, tr_a, tr_b in zip(axs, ne, rt):
    ax.plot(tr_a.times(), tr_a.data,
            label="obspy ({})".format(tr_a.stats.channel))
    ax.plot(tr_b.times(), tr_b.data,
            label="sac ({})".format(tr_b.stats.channel))
    ax.legend()
```

![rotate](https://cloud.githubusercontent.com/assets/77037/20710814/8606e86e-b63c-11e6-802c-150f91b3cd3b.png)

Radial and tangential seem to flip places. If I add 90 degrees to back azimuth, outputs are matched.

```python
ne = obspy.read("./TRIS.IU.BHN.obs.sac")
ne.append(obspy.read("./TRIS.IU.BHE.obs.sac")[0])

rt = obspy.read("./TRIS.IU.BHR.obs.sac")
rt.append(obspy.read("./TRIS.IU.BHT.obs.sac")[0])

baz = rt[0].stats.sac['baz'] + 90

ne.rotate("NE->RT", back_azimuth=baz)

fig, axs = plt.subplots(nrows=2)

for ax, tr_a, tr_b in zip(axs, ne, rt):
    ax.plot(tr_a.times(), tr_a.data,
            label="obspy ({})".format(tr_a.stats.channel))
    ax.plot(tr_b.times(), tr_b.data,
            label="sac ({})".format(tr_b.stats.channel))
    ax.legend()
```
![rotate_plus_90](https://cloud.githubusercontent.com/assets/77037/20710867/d8f4c668-b63c-11e6-9a28-1bbf1e257416.png)

If I use the files with modified sac headers. Outputs are matched again. Because sac output is changed.

```python
ne = obspy.read("./TRIS.IU.BHN.mod.obs.sac")
ne.append(obspy.read("./TRIS.IU.BHE.mod.obs.sac")[0])

rt = obspy.read("./TRIS.IU.BHR.mod.obs.sac")
rt.append(obspy.read("./TRIS.IU.BHT.mod.obs.sac")[0])

baz = rt[0].stats.sac['baz']

ne.rotate("NE->RT", back_azimuth=baz)

fig, axs = plt.subplots(nrows=2)

for ax, tr_a, tr_b in zip(axs, ne, rt):
    ax.plot(tr_a.times(), tr_a.data,
            label="obspy ({})".format(tr_a.stats.channel))
    ax.plot(tr_b.times(), tr_b.data,
            label="sac ({})".format(tr_b.stats.channel))
    ax.legend()
```

![rotate_mod](https://cloud.githubusercontent.com/assets/77037/20710944/49d6bf62-b63d-11e6-94a0-2e61740f0555.png)

Data files I have used: 
[data.zip](https://github.com/obspy/obspy/files/619012/data.zip)

**Comment by QuLogic** [comment]
@jkmacc-LANL

**Comment by megies** [comment]
This is rather a general question, than a SAC specific one I would say. The main reasoning here is that `NE->RT` assumes that input is properly aligned. So, rotating your data to ZNE is another separate step before rotating to ray coordinates.

Please see the longish discussion in #1310, I think that issue covers your problem to a large extent (maybe not the coordinates-from-SAC-header part, but since there is ongoing work you might be able to give valuable feedback on the state of #1310 -- unfortunately I haven't got time to work on it lately, but we should try to get it into the upcoming release)

Also,  a general comment: I would argue that naming a channel with azimuth of 270 degrees `N` is bad practice. Unaligned channels are commonly enumerated, e.g. for borehole instruments, analigned horizontal components are usually named `Z`, `1`, `2`.

CC @krischer

**Comment by jkmacc-LANL** [comment]
From SAC `help rotate`:

>  Both components must be horizontals when the TO option is used.  This
  means that CMPAZ must be definined and that CMPINC must be 90 degrees.  After
  the rotation is completed the first component of each pair will be directed
  along the angle given after to TO keyword.  If the TO GCP option is used this
  component will be directed along the angle given by the station-event back
  azimuth plus or minus 180 degrees.  This component therefore points from the
  event toward the station.  The station and event coordinates header fields
  (STLA, STLO, EVLA, and EVLO) must be defined so that the back azimuth can be
  calculated.

@rdno Right, I think SAC `rotate to gcp` calculates a `rotate to` angle from the `cmpaz` of the horizonals. It seems that ObsPy's `Stream.rotate("NE->RT", back_azimuth=baz)` is more literal; it just selects your 'N' and 'E' components and does something like a `rotate through baz +  180`.  

@megies Is there something the SAC reader can do to store component azimuth/inclination from the SAC header in `Stats` in a useful way?

**Comment by megies** [comment]
> @megies Is there something the SAC reader can do to store component azimuth/inclination from the SAC header in Stats in a useful way?

We do not have a canonical way to store this information in `Stats`, I guess. Maybe we should think about adding some more keys as default well-defined keys in `Stats`..

**Comment by claudiodsf** [comment]
> Maybe we should think about adding some more keys as default well-defined keys in Stats..

I totally agree on this. Another obvious candidate for standardisation is station coordinates.

**Comment by krischer** [comment]
I don't necessarily agree because we already have the inventory objects that store this information. In #1310 we propose to use the inventory objects exactly for this purpose.

Storing this information in the stats header would mainly make it easier for people using the SAC format which is a reasonable goal but IMHO not important enough to modify our internal data model for. I'm not sure how to make it easier for SAC users - maybe add a utility method to transform sac headers to inventory objects?

In any case - I think people using MiniSEED + StationXML are or at least will be at one point (its what the data centers serve after all) more prevalent then people using SAC + RESP files. This is also what our internal data model (`Stream` + `Inventory` objects represents more closely).

**Comment by jkmacc-LANL** [comment]
I'm a SAC user, and I actually agree with @krischer .  I think users can and should put whatever they want to into `Stats`, but I'm not sure that ObsPy should be responsible for it.

**Comment by megies** [comment]
Hmm, yeah, good point there, that would indeed be a step backwards..

One idea: Go after the big catch and make `read` catch-all, returning a three-tuple of `(Stream, Inventory, Catalog)` reading across all different data types (waveforms, station metadata, event metadata) and returning a merge if any information found.. ?

**Comment by rdno** [comment]
> Also, a general comment: I would argue that naming a channel with an azimuth of 270 degrees N is at least very confusing and prone to misinterpretation, if not just plain bad practice. Unaligned channels are commonly enumerated, e.g. for borehole instruments, analigned horizontal components are usually named Z, 1, 2.

I agree it is not a good name for these components. This is how the data came from IRIS.  Maybe, renaming them is more sensible solution. This was such an outlier case I thought it would be a good example. On the other hand, there are some stations with slight variation (e.g. 1 degree azimuth for N component). Considering component azimuth can be  useful for such cases.

Also, SAC does not do the rotation if the components are not orthogonal.  Maybe, that should be checked, as well.

> One idea: Go after the big catch and make read catch-all, returning a three-tuple of (Stream, Inventory, Catalog) reading across all different data types (waveforms, station metadata, event metadata) and returning a merge if any information found.. ?

I think changing the return type of `read` is a big change. How about writing the inventory and event information to stats? If `SAC` file has component and event information, they can be accessed by `tr.stats.inventory` and `tr.stats.catalog`. I also think a function that converts sac headers to inventory object like @krischer suggested can be enough.

I will look at the #1310 and try to give some feedback.

**Comment by claudiodsf** [comment]
I agree with the above idea: if the ObsPy model is `Trace`, `Inventory`, `Catalog`, then reading a SAC trace should populate these three objects, either returning the three, or attaching `Inventory` and `Catalog` to `Trace.stats`.

**Comment by megies** [comment]
Or.. we use the occasion/use case come back to the concept of the `DataSet` object..

```
>>> !ls *
some.mseed
stations.xml
events.xml

>>> data = read("*")

>>> data
DataSet:
  3 Traces
  2 Networks
  3 Events

>>> data.stream  # or data.waveforms?
3 Trace(s) in Stream:
BW.RJOB..EHZ | 2009-08-24T00:20:03.000000Z - 2009-08-24T00:20:32.990000Z | 100.0 Hz, 3000 samples
BW.RJOB..EHN | 2009-08-24T00:20:03.000000Z - 2009-08-24T00:20:32.990000Z | 100.0 Hz, 3000 samples
BW.RJOB..EHE | 2009-08-24T00:20:03.000000Z - 2009-08-24T00:20:32.990000Z | 100.0 Hz, 3000 samples

>>> data.inventory
Inventory created at 2014-03-03T11:07:06.198000Z
	Created by: fdsn-stationxml-converter/1.0.0
		    http://www.iris.edu/fdsnstationconverter
	Sending institution: Erdbebendienst Bayern
	Contains:
		Networks (2):
			BW, GR
		Stations (5):
			BW.RJOB (Jochberg, Bavaria, BW-Net) (3x)
			GR.FUR (Fuerstenfeldbruck, Bavaria, GR-Net)
			GR.WET (Wettzell, Bavaria, GR-Net)
		Channels (30):
			BW.RJOB..EHZ (3x), BW.RJOB..EHN (3x), BW.RJOB..EHE (3x), 
			GR.FUR..BHZ, GR.FUR..BHN, GR.FUR..BHE, GR.FUR..HHZ, GR.FUR..HHN, 
			GR.FUR..HHE, GR.FUR..LHZ, GR.FUR..LHN, GR.FUR..LHE, GR.FUR..VHZ, 
			GR.FUR..VHN, GR.FUR..VHE, GR.WET..BHZ, GR.WET..BHN, GR.WET..BHE, 
			GR.WET..HHZ, GR.WET..HHN, GR.WET..HHE, GR.WET..LHZ, GR.WET..LHN, 
			GR.WET..LHE

>>> data.catalog
3 Event(s) in Catalog:
2012-04-04T14:21:42.300000Z | +41.818,  +79.689 | 4.4 mb | manual
2012-04-04T14:18:37.000000Z | +39.342,  +41.044 | 4.3 ML | manual
2012-04-04T14:08:46.000000Z | +38.017,  +37.736 | 3.0 ML | manual
```

**Comment by aringler-usgs** [comment]
@rdno and @megies Just an FYI: The azimuth of the North being 270 is likely the result of someone installing an STS-2 and accidentally pointing the alignment rod North (it should be pointed East).

TRIS (Tristan da Cunha) is a fairly remote station.  It is possible that the network operator wasn't able to easily change the channel names because of limited communication.  Instead of correcting it, at the risk of data loss, they likely just changed the metadata (a much easier and safer change).

I agree it is definitely not ideal situation, but I would expect to see this from time to time.

**Comment by jkmacc-LANL** [comment]
I *strongly* support the exploration of a `DataSet` or `DataSource` object, which is the space that `Client` objects currently fill.  pyasdf [takes a stab at this](https://seismicdata.github.io/pyasdf/tutorial.html#reading-waveforms-and-stationxml), I don't see a reason that ObsPy couldn't do it.

**Comment by krischer** [comment]
> I agree it is not a good name for these components. This is how the data came from IRIS. Maybe, renaming them is more sensible solution. This was such an outlier case I thought it would be a good example. On the other hand, there are some stations with slight variation (e.g. 1 degree azimuth for N component). Considering component azimuth can be useful for such cases.
> 
> Also, SAC does not do the rotation if the components are not orthogonal. Maybe, that should be checked, as well.

ObsPy does have a function for rotating even non-orthogonal components to ZNE. It just has an awkward interface. We should probably generalize it a bit and make all rotations use that. Its only marginally slower then the other methods but always yields a correct result: http://docs.obspy.org/packages/autogen/obspy.signal.rotate.rotate2zne.html




I also quite strongly think we should go towards a data set object. We cannot change the `read()` function for obvious reasons but we could have a different one. `obspy.gobble("*")` ;-)

The question is how far we want to push this. @megies suggestion would be fairly simple to implement: Just try to read everything with any of the read functions and put it in the corresponding data structure. It becomes quite a bit more powerful once the `DataSet` or whatever we end up calling it is aware of all the different pieces and how they belong together. For waveforms and stations that's quite easy as the SEED identifiers and the times directly define the relations. For events its a bit more difficult. ASDF/pyasdf solve it by enabling the association of waveforms with the QuakeML resource ids. Once all that is available the following would work (and lots of other things that need information about multiple types of data):

```python
ds = obspy.gobble("*")

ds.rotate("->ZRT")  # It knows receiver orientiations and event locations.
ds.remove_response()  # It knows data and instrument responses
...
```

**Comment by jkmacc-LANL** [comment]
> The question is how far we want to push this.

IMO, it's worth doing something major for an ObsPy 2.0, and not really doing much in the 1.X series.

**Comment by claudiodsf** [comment]
I like a lot this idea of `DataSet`!!!
I agree with @jkmacc-LANL that this can be a first major ticket for 2.0.

> We cannot change the read() function for obvious reasons but we could have a different one. obspy.gobble("*") ;-)

I think that in 2.0 we *should* change the `read()` function to what @jkmacc-LANL suggests and we should introduce a `read_waveforms()` function which returns a `Stream` object. This change could be made, say, in ObsPy 1.9 (or 1.2 ;) by deprecating `read()` in favour of `read_waveforms()`.

Maybe is better to further develop this discussion in a new issue...

**Comment by jkmacc-LANL** [comment]
> Maybe is better to further develop this discussion in a new issue...

:+1:

**Comment by claudiodsf** [comment]
> to what @jkmacc-LANL suggests

Sorry, I meant: "to what @megies suggests"

**Comment by jkmacc-LANL** [comment]
Is there an action to take on this specific Issue?  I'm :-1:  on putting a `.inventory` and a `.catalog` on `.stats`, as it would be asking `obspy.io` maintainers to create `Catalog` and `Inventory` objects during their I/O.  I have no philosophical issue with this, it's just a fair amount of work.

**Comment by wjlei1990** [comment]
Hi all,

@rdno bring me to this issue discussion.

So when processing large dataset, we found there are several scenarios:

North | East | Staus
------| -----| -------
0        | 90    | Perfect
5        | 95    | orthogonal but not in alignment
5        | 96    | not orthogonal and bad

What is more, there are also quite a lot of "BH1" and "BH2" components, especially in recent years. I also think if they are orthogonal to each other, they could be used(for example, rotate them to R and T component and they will be fine).

So I write up [my own function](https://github.com/wjlei1990/pytomo3d/blob/master/pytomo3d/signal/rotate.py#L382) to rotate. It checks the `inventory` information to make sure we can use the piece of data:
```
def rotate_one_station_stream(st, event_latitude, event_longitude,
                              station_latitude=None, station_longitude=None,
                              inventory=None, mode="NE->RT",
                              sanity_check=False):
```

**Comment by krischer** [comment]
Its pretty much what we are discussing in #1310 - so your input would of course be very welcome there :-)

Also just a reminder - if you already use your own function do yourself a favor and use this: http://docs.obspy.org/packages/autogen/obspy.signal.rotate.rotate2zne.html

It works for any not linearly dependent three component channel configuration so BH1 and BH2 and what not are fair game.

**Comment by megies** [comment]
I think this still needs some discussion what to do about it, so I'll bump it to 1.2.0
